# 1. ORF length(Open reading frame)

---

open reading frames are defined as spans of DNA sequence between the start and stop codons.

----
**Note :**
  - One common use of open reading frames (ORFs) is as one piece of evidence to assist in gene prediction. Long ORFs are often used, along with other evidence, to initially identify candidate protein-coding regions or functional RNA-coding regions in a DNA sequence.
  - Some short ORFs (sORFs), also named Small open reading frames that lack the classical hallmarks of protein-coding genes (both from ncRNAs and mRNAs) can produce functional peptides.
-----
### Tools:
scikit-learn,TensorFlow, PyTorch, and Keras

----

# Code :
```
from Bio import SeqIO


sequences = [(str([rec.id]) , str(rec.seq)) for rec in SeqIO.parse("test.fasta", "fasta")]

listOfOrf = list()
for index, value in enumerate(sequences):  # looping over the fragments extracted
    frames = [] # storing the six frame translation that it zould be extacted from the fragments
    dna = value[1]  # extract the fragment
    description = value[0] #extact the desciption even were not use it, just for learning purpose
    reverseCdna = [] # storing the reverse compliments
    # create the positive frames
    # split the frames into codons for better performance
    frames.append([dna[i:i + 3] for i in range(0, len(dna), 3)])
    frames.append([dna[i:i + 3] for i in range(1, len(dna), 3)])
    frames.append([dna[i:i + 3] for i in range(2, len(dna), 3)])
    # reverse compliment of the fragment
    reverse = {"A": "T", "C": "G", "T": "A", "G": "C"}
    for i in range(len(dna)):
        reverseCdna.append(reverse[dna[-i - 1]]) if dna[-i - 1] in reverse.keys() else reverseCdna.append(dna[-i - 1])  # if any contamination found we keep it for further more check
    reverseCdna = ''.join(reverseCdna) # joining
    # create the negative frames
    frames.append([reverseCdna[i:i + 3] for i in range(0, len(reverseCdna), 3)])
    frames.append([reverseCdna[i:i + 3] for i in range(1, len(reverseCdna), 3)])
    frames.append([reverseCdna[i:i + 3] for i in range(2, len(reverseCdna), 3)])

```
```
for i in range(0,len(frames),1): #looping all the frames
    start=0
    while start <len(frames[i]): #looping each frame for start and stop codons
        if frames[i][start]=="ATG" or frames[i][start]=="TTG" or frames[i][start]=="CTG" or frames[i][start]=="GTG":
            for stop in range(start+1,len(frames[i]),1):
                         if frames[i][stop]=="TAA" or  frames[i][stop]=="TAG" or  frames[i][stop]=="TGA" :
                                listOfOrf.append(frames[i][start:stop]) # retrieve the orf
                                start=stop+1 # avoiding multiple start codons
                                break
        start+=1





```
```
listOfOrf
```
---

### Sources :

1. [Wikipedia](https://www.pnas.org/doi/10.1073/pnas.1321152111)
2. [ Vlab.amrita](https://vlab.amrita.edu/?sub=3&brch=273&sim=1432&cnt=1)
3. [Open Reading Frame Finder]( https://www.ncbi.nlm.nih.gov/orffinder/)
4. [Great article with code](https://elferachid.medium.com/orf-finder-with-python-3ab35e7ae520)


---
- cool sites didn't read
5. [nih.gov](https://pubmed.ncbi.nlm.nih.gov/29366605/)


