 
# 2. Protein coding potential


Protein coding potential is the probability that a transcript or RNA molecule will be translated into a functional protein.

This is usually determined by analyzing features of the transcript sequence, such as open reading frames, sequence conservation, and other features. Protein coding potential can be assessed using a variety of methods, such as support vector machine-based classifiers , evolutionary modeling, and deep learning algorithms.

---






---

RNA sequencing (RNA-seq) can be used to determine whether a given RNA is coding or non-coding. By sequencing the RNA, it is possible to identify the open reading frames (ORFs) of the transcript, which can then be used to infer the likelihood of the transcript coding for a protein. Additionally, evolutionary conservation, codon usage, and other factors can be taken into account to further assess the protein coding potential of a given transcript.

---


---
 ## Takes a list of sequences and return only the sequences that contain a thioredoxin domain, using the internal HMM:

[https://pyhmmer.readthedocs.io/en/stable/examples/embed_hmms.html](https://pyhmmer.readthedocs.io/en/stable/examples/embed_hmms.html)



---
# Code

```
from Bio.SeqUtils.ProtParam import ProteinAnalysis

my_seq = (
    "MAEGEITTFTALTEKFNLPPGNYKKPKLLYCSNGGHFLRILPDGTVDGTRDRSDQHIQLQ"
    "LSAESVGEVYIKSTETGQYLAMDTSGLLYGSQTPSEECLFLERLEENHYNTYTSKKHAKN"
    "WFVGLKKNGSCKRGPRTHYGQKAILFLPLPV"
)

analysed_seq = ProteinAnalysis(my_seq)

analysed_seq.molecular_weight()

analysed_seq.gravy()

analysed_seq.count_amino_acids()

```
gravy : Calculates the hydrophobicity of amino acid sequences
---


### Source:

1. [HMMER website](http://hmmer.org/documentation.html)
2. [Article by nih to find protein coding](https://pubmed.ncbi.nlm.nih.gov/17631615/)
3. [Pyhammer website](https://pyhmmer.readthedocs.io/en/stable/index.html)
4. [ProtParam](https://biopython.org/wiki/ProtParam)
5. [about .gravy()](https://rdrr.io/cran/alakazam/man/gravy.html)



